File: backend/models/SupportTicket.js
const mongoose = require("mongoose");

const CATEGORY_ENUM = ["ACCOUNT", "BUG", "CONTENT", "OTHER"];
const STATUS_ENUM = ["OPEN", "IN_PROGRESS", "RESOLVED", "CLOSED"];

const supportTicketSchema = new mongoose.Schema(
    {
        user_id: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            required: true,
        },
        subject: {
            type: String,
            required: true,
            trim: true,
            maxlength: [200, "Subject cannot exceed 200 characters"],
        },
        body: {
            type: String,
            required: true,
            trim: true,
            maxlength: [2000, "Body cannot exceed 2000 characters"],
        },
        category: {
            type: String,
            enum: CATEGORY_ENUM,
            default: "OTHER",
        },
        status: {
            type: String,
            enum: STATUS_ENUM,
            default: "OPEN",
        },
        admin_reply: {
            type: String,
            trim: true,
            maxlength: [2000, "Admin reply cannot exceed 2000 characters"],
            default: null,
        },
        replied_at: {
            type: Date,
            default: null,
        },
        replied_by: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            default: null,
        },
    },
    {
        timestamps: true,
        toJSON: {
            transform: (doc, ret) => {
                ret.id = ret._id;
                delete ret._id;
                delete ret.__v;
                return ret;
            },
        },
    },
);

supportTicketSchema.index({ user_id: 1, status: 1 });
supportTicketSchema.index({ status: 1, createdAt: -1 });

const SupportTicket = mongoose.model("SupportTicket", supportTicketSchema);

module.exports = { SupportTicket, CATEGORY_ENUM, STATUS_ENUM };

File: frontend/lib/models/support_ticket.dart
class SupportTicket {
  final String id;
  final String subject;
  final String body;
  final String category;
  final String status;
  final String? adminReply;
  final DateTime? repliedAt;
  final DateTime? createdAt;
  final Map<String, dynamic>? user;

  const SupportTicket({
    required this.id,
    required this.subject,
    required this.body,
    required this.category,
    required this.status,
    this.adminReply,
    this.repliedAt,
    this.createdAt,
    this.user,
  });

  factory SupportTicket.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic v) {
      if (v == null) return null;
      if (v is String) return DateTime.tryParse(v);
      return null;
    }

    final userRaw = json['user_id'];
    final user = userRaw is Map ? Map<String, dynamic>.from(userRaw) : null;

    return SupportTicket(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      subject: (json['subject'] ?? '').toString(),
      body: (json['body'] ?? '').toString(),
      category: (json['category'] ?? 'OTHER').toString(),
      status: (json['status'] ?? 'OPEN').toString(),
      adminReply: json['admin_reply']?.toString(),
      repliedAt: parseDate(json['replied_at']),
      createdAt: parseDate(json['createdAt']),
      user: user,
    );
  }

  bool get isOpen => status == 'OPEN';
  bool get isResolved => status == 'RESOLVED' || status == 'CLOSED';
  bool get hasReply => adminReply != null && adminReply!.isNotEmpty;
}
